### PR TITLE
time: fix leap year maths in datetime_to_time

### LINKED ISSUE #3235: datetime_to_time reports incorrect date
## Context

* Operating System & Odin Version: Linux 6.7.6, Odin commit 7df7fec6f736a52712961a2a54de3747eac4ec84

## Expected Behavior

The following program should output a date of March 1st, but it reports February the 29th.

```odin
package main

import "core:fmt"
import "core:time"

main :: proc() {
	fmt.println(time.datetime_to_time(2024, 3, 1, 0, 0, 0))
}
```

## Failure Information (for bugs)

The program prints this:
```
2024-02-29 00:00:00.000000000 +0000 UTC true
```

I think the procedure isn't taking into account leap years with regard to February's extra day, and the logic behind `DAYS_PER_400_YEARS` and the other calculations is sound, but it's starting the arithmetic modulo from the year minus 1970, which would give incorrect results.

### PR DESCRIPTION
**Pull request by widowight**
I caught a bug with `datetime_to_time` where it was failing to calculate days according to leap years.
Also replaced the internal `is_leap_year` proc with a public API that uses bitwise instead of modulo calculations.

I calculated `EPOCH_DAYS` with this program:

```odin
	for i in 1 ..< 1970 {
		days += 365

		if i > 0 && is_leap_year(i) {
			days += 1
		}
	}
```

Fixes #3235

### CODE REVIEW

File: core/time/time.odin [hunk 1]
```diff
@@ -360,7 +356,13 @@ _abs_date :: proc "contextless" (abs: u64, full: bool) -> (year: int, month: Mon
 	return
 }
 
+is_leap_year :: proc "contextless" (year: int) -> bool {
+	return !((year & 3 > 0) || (year & 5 > 0) && !(year % 25 == 0))
+}
```
Comment (gingerBill) [inline core/time/time.odin:361]: Why did you remove the clearer logic and replace it with "magic" (I understand what it does)?

**Comment by gingerBill** [comment]
Fixed elsewhere.

File: core/time/time.odin (hunk 1 above)
Comment (widowight) [reply core/time/time.odin:361]: An understanding that partially bitwise arithmetic would be faster than completely modular arithmetic.
